linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/JSBuilders.scala

```diff
@@ -49,24 +52,24 @@ class JSFileBuilder(val name: String,
     // Do not close the printer: we do not have ownership of the writers
   }
 
-  /** Closes the underlying writer(s).
-   */
-  def closeWriters(): Unit = {
+  def complete(): Unit =
     outputWriter.close()
-  }
 }
 
-class JSFileBuilderWithSourceMapWriter(n: String, ow: Writer,
-    protected val sourceMapWriter: SourceMapWriter)
-    extends JSFileBuilder(n, ow) {
+class JSFileBuilderWithSourceMap(name: String, outputWriter: Writer,
+    sourceMapOutputWriter: Writer,
+    relativizeSourceMapBasePath: Option[URI]) extends JSLineBuilder {
+  private val sourceMapWriter = new SourceMapWriter(
+      sourceMapOutputWriter, name, relativizeSourceMapBasePath)
 
-  override def addLine(line: String): Unit = {
-    super.addLine(line)
+  def addLine(line: String): Unit = {
+    outputWriter.write(line)
+    outputWriter.write('\n')
     sourceMapWriter.nextLine()
   }
 
-  override def addTopLevelStatement(originalLocation: URI, code: String): Unit = {
-    super.addTopLevelStatement(originalLocation, code)
+  def addTopLevelStatement(originalLocation: URI, code: String): Unit = {
+    outputWriter.write(code)
```

Comment:
Ah, so with this commit, there is no need for two methods to stay in sync, which is good. But the logic in this method must still send a `\n` to the `outputWriter` if `code` does not have a trailing `\n`.